### DISCUSSION TITLE
unable to run tests with maven

### DISCUSSION THREAD

**Discussion by kikkauz** [dasniko/testcontainers-keycloak#104] [Q&A]
I'm setting up tests using the KeycloakContainer to test a custom provider but running the test goal in maven fails with an error:
java.lang.IllegalStateException: Mapped port can only be obtained after the container is started
	at org.testcontainers.shaded.com.google.common.base.Preconditions.checkState(Preconditions.java:174)
	at org.testcontainers.containers.ContainerState.getMappedPort(ContainerState.java:149)
	at dasniko.testcontainers.keycloak.ExtendableKeycloakContainer.getHttpPort(ExtendableKeycloakContainer.java:398)
	at dasniko.testcontainers.keycloak.ExtendableKeycloakContainer.getAuthServerUrl(ExtendableKeycloakContainer.java:386)
	at org.iso.keycloak.custom.federation.utils.FederationUtilsProviderTest.givenSpec(FederationUtilsProviderTest.java:91)
	at org.iso.keycloak.custom.federation.utils.FederationUtilsProviderTest.testQueryUnauthenticatedUser(FederationUtilsProviderTest.java:32)

The issue is that it seems the test container is not started when test are executed (I need to test custom rest endpoints).

If I start the tests within intelliJ it works just fine:
.....
[main] INFO org.testcontainers.utility.RyukResourceReaper - Ryuk started - will monitor and terminate Testcontainers containers on JVM exit
[main] INFO org.testcontainers.DockerClientFactory - Checking the system...
[main] INFO org.testcontainers.DockerClientFactory - ✔︎ Docker server version should be at least 1.6.0
[main] INFO 🐳 [quay.io/keycloak/keycloak:21.0] - Creating container for image: quay.io/keycloak/keycloak:21.0
....

while executing tests with maven, it does not even try to create the keycloak container and executes directly tests.

I hope someone can have hints for me.
Thanks!

**Comment by kikkauz** [comment]
Not sure why, but I had to explicitly start the container with the .start() method.

**Reply by dasniko** [reply]
As you don't show us your code, we can hardly tell you, what might probably be the cause. 🤷‍♂️

You might want to have a look here, where I run my realm-resource tests: https://github.com/dasniko/keycloak-extensions-demo/blob/main/rest-endpoint/src/test/java/dasniko/keycloak/resource/MyResourceProviderTest.java

**Reply by kikkauz** [reply]
Hello. Thanks for the reply. You are absolutely right, I forgot to mention that I'm taking inspiration from your code you just linked. 
So basically, to make tests working when executed by maven, I had to change the method givenSpec() as
if (!keycloak.isRunning()) {
      keycloak.start();
    }

    return given()
        .baseUri(keycloak.getAuthServerUrl())
        .basePath("/realms/master/" + FederationUtilsProviderFactory.PROVIDER_ID)
        .sessionId((new Random()).toString());
